File: dashboard-backend/dashboard/src/main/java/com/mediahost/dashboard/model/entity/SiteMonitor.java
package com.mediahost.dashboard.model.entity;

import com.mediahost.dashboard.model.enums.*;
import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.Data;
import java.time.LocalDateTime;
import java.time.LocalTime;

@Entity
@Table(name = "site_monitors")
@Data
public class SiteMonitor {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @NotBlank(message = "Name is required")
    @Size(min = 3, max = 200)
    @Column(length = 200, nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotBlank(message = "URL is required")
    @Pattern(regexp = "^(http|https)://.*$", message = "Invalid URL format")
    @Column(length = 500, nullable = false)
    private String url;

    @Column(name = "expected_phrase", length = 255)
    private String expectedPhrase;

    @Column(name = "expected_phrase_missing")
    private Boolean expectedPhraseMissing = false;

    @Column(name = "retry_count")
    @Min(0) @Max(10)
    private Integer retryCount = 3;

    @Column(name = "retry_interval_seconds")
    @Min(5) @Max(3600)
    private Integer retryIntervalSeconds = 60;

    @Column(name = "check_interval_minutes")
    @Min(1) @Max(1440)
    private Integer checkIntervalMinutes = 5;

    @Column(name = "timeout_seconds")
    @Min(1) @Max(120)
    private Integer timeoutSeconds = 30;

    @Column(name = "follow_redirects")
    private Boolean followRedirects = true;

    @Column(name = "expected_status_code")
    @Min(100) @Max(599)
    private Integer expectedStatusCode = 200;

    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false)
    private ServiceType serviceType;

    @Enumerated(EnumType.STRING)
    private Severity severity = Severity.medium;

    @Column(name = "business_hours_only")
    private Boolean businessHoursOnly = false;

    @Column(name = "business_hours_start")
    private LocalTime businessHoursStart = LocalTime.of(6, 0);

    @Column(name = "business_hours_end")
    private LocalTime businessHoursEnd = LocalTime.of(20, 0);

    @Column(name = "business_days", length = 50)
    private String businessDays = "1,2,3,4,5";

    @Column(name = "is_active")
    private Boolean isActive = true;

    private Boolean alert = true;

    @Column(name = "alert_email")
    @Email
    private String alertEmail;

    @Column(name = "last_check_time")
    private LocalDateTime lastCheckTime;

    @Enumerated(EnumType.STRING)
    @Column(name = "last_check_status")
    private MonitorStatus lastCheckStatus;

    @Column(name = "last_check_message", columnDefinition = "TEXT")
    private String lastCheckMessage;

    @Column(name = "current_failure_count")
    private Integer currentFailureCount = 0;

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "updated_by")
    private Integer updatedBy;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: dashboard-frontend/src/app/core/services/site-monitor.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface SiteMonitor {
  id?: number;
  name: string;
  description?: string;
  url: string;
  expectedPhrase?: string;
  expectedPhraseMissing?: boolean;
  retryCount: number;
  retryIntervalSeconds: number;
  checkIntervalMinutes: number;
  timeoutSeconds: number;
  followRedirects?: boolean;
  expectedStatusCode: number;
  serviceType: 'Alerts' | 'Broadcast' | 'Deliveries' | 'Online' | 'Print';
  severity: 'critical' | 'high' | 'medium' | 'low';
  businessHoursOnly?: boolean;
  businessHoursStart?: string;
  businessHoursEnd?: string;
  businessDays?: string;
  isActive?: boolean;
  alert?: boolean;
  alertEmail?: string;
  lastCheckTime?: string;
  lastCheckStatus?: string;
  lastCheckMessage?: string;
  currentFailureCount?: number;
}

@Injectable({
  providedIn: 'root'
})
export class SiteMonitorService {
  private apiUrl = 'http://localhost:8080/api/monitors/sites';

  constructor(private http: HttpClient) {}

  getAll(): Observable<SiteMonitor[]> {
    return this.http.get<SiteMonitor[]>(this.apiUrl);
  }

  getById(id: number): Observable<SiteMonitor> {
    return this.http.get<SiteMonitor>(`${this.apiUrl}/${id}`);
  }

  create(data: SiteMonitor): Observable<SiteMonitor> {
    return this.http.post<SiteMonitor>(this.apiUrl, data);
  }

  update(id: number, data: SiteMonitor): Observable<SiteMonitor> {
    return this.http.put<SiteMonitor>(`${this.apiUrl}/${id}`, data);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  resetFailureCount(id: number): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}/reset`, {});
  }

  getFailing(): Observable<SiteMonitor[]> {
    return this.http.get<SiteMonitor[]>(`${this.apiUrl}/failing`);
  }

  getAlerting(): Observable<SiteMonitor[]> {
    return this.http.get<SiteMonitor[]>(`${this.apiUrl}/alerting`);
  }

  recheck(id: number): Observable<SiteMonitor> {
    return this.http.post<SiteMonitor>(`${this.apiUrl}/${id}/recheck`, {});
  }
}